rust/lance-io/src/object_store/providers.rs

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::{collections::HashMap, sync::Arc};
+
+use snafu::location;
+use url::Url;
+
+use super::{ObjectStore, ObjectStoreParams};
+use lance_core::error::{Error, Result};
+
+#[cfg(feature = "aws")]
+pub mod aws;
+#[cfg(feature = "azure")]
+pub mod azure;
+#[cfg(feature = "gcp")]
+pub mod gcp;
+pub mod local;
+pub mod memory;
+
+#[async_trait::async_trait]
+pub trait ObjectStoreProvider: std::fmt::Debug + Sync + Send {
+    async fn new_store(&self, base_path: Url, params: &ObjectStoreParams) -> Result<ObjectStore>;
+}
+
+#[derive(Debug)]
+pub struct ObjectStoreRegistry {
+    providers: HashMap<String, Arc<dyn ObjectStoreProvider>>,
+    // TODO: should the registry itself have a cache?
```

Comment:
If we have one instance of an object store I wonder if we can, later, move some of the IOPS-throttling to the object store.  Right now each scan has its own throttle but it might be nicer to have a process-wide throttle.

Reply:
Yeah we might consider it. We still might have duplicate stores as the cache will likely be keyed based on the `ObjectStoreParams` / `storage_options`. So if a user opens each table with slightly different settings, they will have different stores. Still, better than one-per scan.